tests/unit/test_hnsw_tiered.cpp

```diff
@@ -436,3 +436,465 @@ TYPED_TEST(HNSWTieredIndexTest, insertJobAsyncMulti) {
+    for (auto isMulti : {false, true}) {
+        HNSWParams params = {
+            .type = TypeParam::get_index_type(),
+            .dim = dim,
+            .metric = VecSimMetric_L2,
+            .multi = isMulti,
+            .efRuntime = 100,
+        };
+        auto jobQ = JobQueue();
+        auto index_ctx = IndexExtCtx();
+        size_t memory_ctx = 0;
+        TieredIndexParams tiered_params = {
+            .jobQueue = &jobQ,
+            .jobQueueCtx = &index_ctx,
+            .submitCb = submit_callback,
+            .memoryCtx = &memory_ctx,
+            .UpdateMemCb = update_mem_callback,
+        };
+        TieredHNSWParams tiered_hnsw_params = {.hnswParams = params, .tieredParams = tiered_params};
+        auto *tiered_index = reinterpret_cast<TieredHNSWIndex<TEST_DATA_T, TEST_DIST_T> *>(
+            HNSWFactory::NewTieredIndex(&tiered_hnsw_params, allocator));
+        // Set the created tiered index in the index external context.
+        index_ctx.index_strong_ref.reset(tiered_index);
+        EXPECT_EQ(index_ctx.index_strong_ref.use_count(), 1);
+
+        std::atomic_int successful_searches(0);
+        auto parallel_knn_search = [](AsyncJob *job) {
+            auto *search_job = reinterpret_cast<SearchJobMock *>(job);
+            size_t k = search_job->k;
+            size_t dim = search_job->dim;
+            auto query = search_job->query;
+
+            auto verify_res = [&](size_t id, double score, size_t res_index) {
+                TEST_DATA_T el = *(TEST_DATA_T *)query;
+                ASSERT_EQ(std::abs(id - el), (res_index + 1) / 2);
+                ASSERT_EQ(score, dim * (id - el) * (id - el));
+            };
+            runTopKSearchTest(job->index, query, k, verify_res);
+            search_job->successful_searches++;
+
+            delete search_job;
+        };
+
+        size_t per_label = isMulti ? 10 : 1;
+        size_t n_labels = n / per_label;
+
+        // Fill the job queue with insert and search jobs, while filling the flat index, before
+        // initializing the thread pool.
+        for (size_t i = 0; i < n; i++) {
+            // Insert a vector to the flat index and add a job to insert it to the main index.
+            GenerateAndAddVector<TEST_DATA_T>(tiered_index, dim, i % n_labels, i);
+
+            // Add a search job. Make sure the query element is between k and n - k.
+            auto query = (TEST_DATA_T *)allocator->allocate(dim * sizeof(TEST_DATA_T));
+            GenerateVector<TEST_DATA_T>(query, dim, (i % (n_labels - (2 * k))) + k);
+            auto search_job =
+                new (allocator) SearchJobMock(allocator, parallel_knn_search, tiered_index, query,
+                                              k, n, dim, successful_searches);
+            tiered_index->submitSingleJob(search_job);
+        }
+
+        EXPECT_EQ(tiered_index->indexSize(), n) << (isMulti ? "multi" : "single");
```

Comment:
think about better debug print?

Reply:
I made one in the next PR, I can move it to this one